File: 63-unique-paths-ii/63-unique-paths-ii.java
class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        int m = obstacleGrid.length;
        int n = obstacleGrid[0].length;
        int[] dp = new int[n];
        dp[0] = 1;
        
        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                if (j > 0) {
                    dp[j] += dp[j - 1];
                }
                dp[j] *= (1 - obstacleGrid[i][j]);
            }
        }
        return dp[n - 1];
    }
}

File: 63-unique-paths-ii/63-unique-paths-ii.js
/**
 * @param {number[][]} obstacleGrid
 * @return {number}
 */
var uniquePathsWithObstacles = function(obstacleGrid) {
    const n = obstacleGrid.length
    const m = obstacleGrid[0].length;
    let prev = new Array(m + 1).fill(0);
    let curr = new Array(m + 1).fill(0);
    curr[0] = 1;
    
    for(let i = 0; i < n; i++) {
        for(let j = 1; j <= m; j++) {
            if (!obstacleGrid[i][j-1]){
                curr[j] = prev[j] + curr[j - 1];
            }
        }
        prev = curr;
        curr = new Array(m + 1).fill(0);
    }
    
    return prev[m];
};